File: cpp/se/Greeting.cpp
#include "se/Greeting.hpp"
#include "se/EPPDateFormatter.hpp"
#include "xml/ParsingException.hpp"

const std::string Greeting::GREETING_EXPR ("/e:epp/e:greeting");
const std::string Greeting::DCP_EXPR (GREETING_EXPR + "/e:dcp");
const std::string Greeting::SVID_EXPR(GREETING_EXPR + "/e:svID/text()");
const std::string Greeting::SVDATE_EXPR(GREETING_EXPR + "/e:svDate/text()");
const std::string Greeting::VERSIONS_EXPR (GREETING_EXPR + "/e:svcMenu/e:version");
const std::string Greeting::LANGS_EXPR (GREETING_EXPR + "/e:svcMenu/e:lang");
const std::string Greeting::OBJ_URIS_EXPR (GREETING_EXPR + "/e:svcMenu/e:objURI");
const std::string Greeting::EXT_URIS_EXPR (GREETING_EXPR + "/e:svcMenu/e:svcExtension/e:extURI");
const std::string Greeting::ACCESS_EXPR (DCP_EXPR + "/e:access/*[1]");
const std::string Greeting::STMT_COUNT_EXPR ("count(" + DCP_EXPR + "/e:statement)");
const std::string Greeting::STMT_IND_EXPR (DCP_EXPR + "/e:statement[IDX]");
const std::string Greeting::PURPOSE_EXPR ("/e:purpose");
const std::string Greeting::RECIPIENT_EXPR ("/e:recipient");
const std::string Greeting::RETENTION_EXPR ("/e:retention/*[1]");
const std::string Greeting::EXPIRY_EXPR ("/e:expiry/*[1]");


std::string Greeting::toString() const
{
    unsigned int i;
    std::string versionString = versions[0];
    
    for (i = 1; i < versions.size(); i++)
        versionString += "," + versions[i];
    
    std::string langString = langs[0];
    for (i = 1; i < langs.size(); i++)
        langString += "," + langs[i];
    
    std::string objURIString = objURIs[0];
    for (i = 1; i < objURIs.size(); i++)
        objURIString += "," + objURIs[i];
    
    std::string retval = 
        "(svID = " + getServerID() + ")" +
        (getServerDateTime() != NULL ? "(svDate = " + getServerDateTime()->toXMLFormat() + ")" : "") +
        "(versions = (" + versionString + "))" +
        "(languages = (" + langString + "))" +
        "(objURIs = (" + objURIString + "))";
    
    return retval;
}


void Greeting::fromXML(XMLDocument *xmlDoc) throw (ParsingException)
{
    debugLogger->LOG_FINEST("enter");
    
    try
    {
		// debugLogger->info (xmlDoc->toString());
        svID = xmlDoc->getNodeValue (SVID_EXPR);
        std::string svDateText = xmlDoc->getNodeValue (SVDATE_EXPR);
        svDate = std::auto_ptr<XMLGregorianCalendar>(EPPDateFormatter::fromXSDateTime(svDateText));
        versions = xmlDoc->getNodeValues(VERSIONS_EXPR);
        langs = xmlDoc->getNodeValues(LANGS_EXPR);
        objURIs = xmlDoc->getNodeValues(OBJ_URIS_EXPR);
        extURIs = xmlDoc->getNodeValues(EXT_URIS_EXPR);
        dcpAccess = xmlDoc->getNodeValue(ACCESS_EXPR);
        
        int dcpStmtCount = xmlDoc->getNodeCount(STMT_COUNT_EXPR);
        for (int i = 0; i < dcpStmtCount; i++)
        {
            std::string qry = ReceiveSE::replaceIndex(STMT_IND_EXPR, i+1);

            dcpStatements.push_back(
				DCPStatement(
					xmlDoc->getChildNames(qry + PURPOSE_EXPR),
					xmlDoc->getChildNames(qry + RECIPIENT_EXPR),
					xmlDoc->getNodeName(qry + RETENTION_EXPR)));
        }
    }
    catch (XPathExpressionException& e)
	{
		ParsingException pe;
		pe.causedBy(e);
		throw pe;
	}
    
    debugLogger->info (toString());
    debugLogger->LOG_FINEST("exit");
}

File: src/main/java/com/ausregistry/jtoolkit2/se/Greeting.java
package com.ausregistry.jtoolkit2.se;

import java.io.Serializable;
import java.util.GregorianCalendar;

import javax.xml.xpath.XPathExpressionException;

import com.ausregistry.jtoolkit2.EPPDateFormatter;
import com.ausregistry.jtoolkit2.xml.XMLDocument;

/**
 * Provides access to EPP server parameters as published in an EPP greeting
 * service element.
 *
 */
public final class Greeting extends ReceiveSE {
    private static final long serialVersionUID = 4358903746268469400L;

    private static final String GREETING_EXPR = "/e:epp/e:greeting";
    private static final String DCP_EXPR = GREETING_EXPR + "/e:dcp";
    private static final String SVID_EXPR = GREETING_EXPR + "/e:svID/text()";
    private static final String SVDATE_EXPR = GREETING_EXPR + "/e:svDate/text()";
    private static final String VERSIONS_EXPR = GREETING_EXPR + "/e:svcMenu/e:version";
    private static final String LANGS_EXPR = GREETING_EXPR + "/e:svcMenu/e:lang";
    private static final String OBJ_URIS_EXPR = GREETING_EXPR + "/e:svcMenu/e:objURI";
    private static final String EXT_URIS_EXPR = GREETING_EXPR + "/e:svcMenu/e:svcExtension/e:extURI";
    private static final String ACCESS_EXPR = DCP_EXPR + "/e:access/*[1]";
    private static final String EXPIRY_EXPR = DCP_EXPR + "/e:expiry/*[1]";
    private static final String STMT_COUNT_EXPR = "count(" + DCP_EXPR + "/e:statement)";
    private static final String STMT_IND_EXPR = DCP_EXPR + "/e:statement[IDX]";
    private static final String PURPOSE_EXPR = "/e:purpose";
    private static final String RECIPIENT_EXPR = "/e:recipient";
    private static final String RETENTION_EXPR = "/e:retention/*[1]";

    private String svID;
    private GregorianCalendar svDate;
    private String[] versions;
    private String[] langs;
    private String[] objURIs;
    private String[] extURIs;
    private String dcpAccess;
    private String dcpExpiry;
    private DCPStatement[] dcpStatements;

    public Greeting() {
        super();
    }

    public String getServerID() {
        return svID;
    }

    public GregorianCalendar getServerDateTime() {
        return svDate;
    }

    public String[] getProtocolVersions() {
        return versions;
    }

    public String[] getLanguages() {
        return langs;
    }

    public String[] objURIs() {
        return objURIs;
    }

    public String[] extURIs() {
        return extURIs;
    }

    public String dcpAccess() {
        return dcpAccess;
    }

    public String dcpExpiry() {
        // Known limitation: actual expiry date or period is not available.
        return dcpExpiry;
    }

    public DCPStatement[] getDataCollectionPolicyStatements() {
        return dcpStatements;
    }

    @Override
    public String toString() {
        String versionString = arrayToString(versions, ",");
        String langString = arrayToString(langs, ",");
        String objURIString = arrayToString(objURIs, ",");

        String retval = "(svID = " + getServerID() + ")" + "(svDate = "
                + getServerDateTime() + ")" + "(versions = (" + versionString
                + "))" + "(languages = (" + langString + "))" + "(objURIs = ("
                + objURIString + "))";

        return retval;
    }

    @Override
    public void fromXML(XMLDocument xmlDoc) {
        debugLogger.finest("enter");

        try {
            debugLogger.info(xmlDoc.toString());
            svID = xmlDoc.getNodeValue(SVID_EXPR);
            String svDateText = xmlDoc.getNodeValue(SVDATE_EXPR);
            svDate = EPPDateFormatter.fromXSDateTime(svDateText);
            versions = xmlDoc.getNodeValues(VERSIONS_EXPR);
            langs = xmlDoc.getNodeValues(LANGS_EXPR);
            objURIs = xmlDoc.getNodeValues(OBJ_URIS_EXPR);
            extURIs = xmlDoc.getNodeValues(EXT_URIS_EXPR);
            dcpAccess = xmlDoc.getNodeName(ACCESS_EXPR);
            dcpExpiry = xmlDoc.getNodeName(EXPIRY_EXPR);

            int dcpStmtCount = xmlDoc.getNodeCount(STMT_COUNT_EXPR);
            dcpStatements = new DCPStatement[dcpStmtCount];
            for (int i = 0; i < dcpStmtCount; i++) {
                String qry = ReceiveSE.replaceIndex(STMT_IND_EXPR, i + 1);
                dcpStatements[i] = new DCPStatement(xmlDoc.getChildNames(qry
                        + PURPOSE_EXPR), xmlDoc.getChildNames(qry
                        + RECIPIENT_EXPR), xmlDoc.getNodeName(qry
                        + RETENTION_EXPR));
            }
        } catch (XPathExpressionException xpee) {
            maintLogger.warning(xpee.getMessage());
        }

        debugLogger.info(toString());
        debugLogger.finest("exit");
    }

    /**
     * Provides access to the data collection policy statement parameter values
     * reported by an EPP server in a Greeting compliant with the greeting
     * specification in RFC5730.
     */
    public static final class DCPStatement implements Serializable {
        private static final long serialVersionUID = -589856890805744448L;
        private String[] purposes;
        private String[] recipients;
        private String retention;

        DCPStatement(String[] purposes, String[] recipients,
                String retentionPolicy) {

            this.purposes = purposes;
            this.recipients = recipients;
            this.retention = retentionPolicy;
        }

        public String[] getPurpose() {
            return purposes;
        }

        public String[] getRecipients() {
            // Known limitation: the recDesc child of 'ours' is not recorded.
            return recipients;
        }

        public String getRetentionPolicy() {
            return retention;
        }
    }
}
